File: mirea/cources/beginner/connected-components-3/f.py
# структура для системы непересекающихся множеств
class DSU:
    
    def __init__(self, n):
        self.sz = [1] * n;
        self.par = [int(x) for x in range(n)];
        self.numEdges = [0] * n;
    
    def getPar(self, u):
        if u == self.par[u]:
            return u;
        p = self.getPar(self.par[u]);
        self.par[u] = p;
        return p;
    
    def connect(self, u, v):
        u, v = self.getPar(u), self.getPar(v);
        if u != v:
            if self.sz[u] < self.sz[v]:
                u, v = v, u; # swap
            self.sz[u] += self.sz[v];
            self.par[v] = u;
            self.numEdges[u] += self.numEdges[v];
        self.numEdges[u] += 1;

# решение задачи
# те вершины, которые соединены рёбрами, объединяем в компоненты
# связности, считая при этом количество рёбер в каждой компоненте
# (подсчёт ведётся внутри DSU)
nVert, nEdges = (int(x) for x in input().split())
dsu = DSU(nVert)
for i in range(nEdges):
    u, v = (int(x)-1 for x in input().split())
    dsu.connect(u, v);
# если компонента является циклом (кол-во рёбер равно колву вершин)
# или в ней есть цикл (ребёр больше чем вершин)
# то для неё ответ равен 0: можно по циклу направить рёбра,
# иначе эта компонента является деревом и в этой компоненте ответ
# равен 1, так как будет ровно одна вершина,
# в которую не будет вести ни одно ребро (рёбер на 1 меньше чем вершин)

# отмечаем, какие вершины являются представителями компонент связностей
isParent = [False] * nVert
for i in range(nVert):
    isParent[dsu.getPar(i)] = True
# считаем ответ: количество деревьев
answ = 0
for p in range(nVert):
    if isParent[p]:
        answ += max(0, (dsu.sz[p] - dsu.numEdges[p]))
print(answ)

File: mirea/cources/beginner/connected-components-3/f.cpp
#include <iostream>
#include <vector>
using namespace std;

// Структура для системы непересекающихся множеств
class DSU {
public:
    vector<int> sz, par, numEdges;

    DSU(int n) {
        sz.resize(n, 1);
        par.resize(n);
        numEdges.resize(n, 0);
        for (int i = 0; i < n; i++) {
            par[i] = i;
        }
    }

    int getPar(int u) {
        if (u == par[u]) return u;
        return par[u] = getPar(par[u]); // Сжатие пути
    }

    void connect(int u, int v) {
        u = getPar(u);
        v = getPar(v);
        if (u != v) {
            if (sz[u] < sz[v]) swap(u, v);
            sz[u] += sz[v];
            par[v] = u;
            numEdges[u] += numEdges[v];
        }
        numEdges[u] += 1;
    }
};

int main() {
    // Ускорение ввода-вывода
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int nVert, nEdges;
    cin >> nVert >> nEdges;
    DSU dsu(nVert);
    
    for (int i = 0; i < nEdges; i++) {
        int u, v;
        cin >> u >> v;
        dsu.connect(u - 1, v - 1);
    }
    
    // Отмечаем, какие вершины являются представителями компонент связностей
    vector<bool> isParent(nVert, false);
    for (int i = 0; i < nVert; i++) {
        isParent[dsu.getPar(i)] = true;
    }
    
    // Подсчет количества деревьев
    int answ = 0;
    for (int p = 0; p < nVert; p++) {
        if (isParent[p]) {
            answ += max(0, dsu.sz[p] - dsu.numEdges[p]);
        }
    }
    
    cout << answ << "\n";
    return 0;
}
